mux.js

```diff
@@ -0,0 +1,248 @@
+var protocol = require('hypercore-protocol')
+var readify = require('./ready')
+var inherits = require('inherits')
+var events = require('events')
+var debug = require('debug')('multifeed')
+var hypercore = require('hypercore')
+var xtend = require('xtend')
+
+// constants
+var MULTIFEED = 'MULTIFEED'
+var PROTOCOL_VERSION = '2.0.0'
+// extensions
+var MANIFEST = 'MANIFEST'
+var REQUEST_FEEDS = 'REQUEST_FEEDS'
+/*
+var ANNOUNCE_FEED = 'ANNOUNCE_FEED'
+var REQUEST_FEED_SIGNATURE = 'REQUEST_FEED_SIGNATURE'
+var FEED_SIGNATURE = 'FEED_SIGNATURE'
+*/
+
+var SupportedExtensions = [
+  MANIFEST,
+  REQUEST_FEEDS
+  //ANNOUNCE_FEED
+  //REQUEST_MANIFEST,
+  //REQUEST_FEED_SIGNATURE,
+  //FEED_SIGNATURE,
+]
+
+// `key` - protocol encryption key
+// `opts`- hypercore-protocol opts
+function Multiplexer (key, opts) {
+  if (!(this instanceof Multiplexer)) return new Multiplexer(key, opts)
+  debug('[REPLICATION] New mux initialized', key.toString('hex'), opts)
+  var self = this
+  self._opts = opts = opts || {}
+  self.extensions = opts.extensions = SupportedExtensions || opts.extensions
+
+  // initialize
+  self._localHave = null
+  self._localWant = null
+  self._remoteHas = null
+  self._remoteWants = null
+
+  var stream = this.stream = protocol(Object.assign(opts,{
+    userData: Buffer.from(JSON.stringify({
+      client: MULTIFEED,
+      version: PROTOCOL_VERSION,
+      // Help! exposing available extensions might be good for future version tolerance,
+      // but at the same time we're poentiallyleaking our user-agent fingerprint to a third party.
+      extensions: self.extensions
+    }))
+  }))
+
+  var feed = this._feed = stream.feed(key)
+
+  stream.on('handshake', function () {
+    var header = JSON.parse(this.userData.toString('utf8'))
+    debug('[REPLICATION] recv\'d header: ', JSON.stringify(header))
+    if (!compatibleVersions(header.version, PROTOCOL_VERSION)) {
+      debug('[REPLICATION] aborting; version mismatch (us='+PROTOCOL_VERSION+')')
+      self._finalize(new Error('protocol version mismatch! us='+PROTOCOL_VERSION + ' them=' + header.version))
+      return
+    }
+
+    if (header.client != MULTIFEED) {
+      debug('[REPLICATION] aborting; Client mismatch! expected ', MULTIFEED, 'but got', header.client)
+      self._finalize(new Error('Client mismatch! expected ' + MULTIFEED + ' but got ' + header.client))
+      return
+    }
+    self.remoteClient = header
+    self.emit('ready', header)
+  })
+
+  feed.on('extension', function (type, message) {
+    debug('Extension:', type, message.toString('utf8'))
+    switch(type) {
+      case MANIFEST:
+        var rm = JSON.parse(message.toString('utf8'))
+        self._remoteHas = rm.keys
+        self.emit('manifest', rm)
+        break
+      case REQUEST_FEEDS:
+        self._remoteWants = JSON.parse(message.toString('utf8'))
+        self._initRepl()
+        break
+      // case ANNOUNCE_FEED:
+      //  self._announceFeed(JSON.parse(message.toString('utf8')))
```

Comment:
dead code :ghost:

Reply:
Uh this is a mental note to implement 'new feed available' during live-replication.
To mitigate the fact that the feed-exchange only occurs during new connections.

The code is untested but it should be the starting point of doing a second feed-exchange when a new feed becomes available at a node and we want to announce it's existence to other connected nodes without having to reconnect.

Reply:
Yeah, I think keeping untested commented code around is confusing. Maybe keep this code, uncommented, in a feature branch so that it's still there for you for future dev?

Reply:
When you put it that way. Yeah I agree.  I had optimistically hoped that someone wanted to take over it and finish that feature but it's better to move those lines into a new Issue & Featureand remove them from the stable branch as you suggested.